metadata/src/main/java/org/apache/kafka/controller/DelegationTokenControlManager.java

```diff
@@ -0,0 +1,359 @@
+                .setPrincipalName(owner.getName())
+                .setPrincipalType(owner.getPrincipalType())
+                .setTokenRequesterPrincipalName(context.principal().getName())
+                .setTokenRequesterPrincipalType(context.principal().getPrincipalType());
+
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+
+        if (secretKeyString == null) {
+            // DelegationTokens are not enabled
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_AUTH_DISABLED.code()));
+        }
+
+        if (!metadataVersion.isDelegationTokenSupported()) {
+            // DelegationTokens are not supported in this metadata version
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(UNSUPPORTED_VERSION.code()));
+        }
+
+        List<KafkaPrincipal> renewers = new ArrayList<KafkaPrincipal>();
+        for (CreatableRenewers renewer : requestData.renewers()) {
+            if (renewer.principalType().equals(KafkaPrincipal.USER_TYPE)) {
+                renewers.add(new KafkaPrincipal(renewer.principalType(), renewer.principalName()));
+            } else {
+                return ControllerResult.atomicOf(records, responseData.setErrorCode(INVALID_PRINCIPAL_TYPE.code()));
+            }
+        }
+
+        byte[] hmac;
+        try {
+            hmac = createHmac(tokenId);
+        } catch (Throwable e) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(ApiError.fromThrowable(e).error().code()));
+        }
+
+        TokenInformation newTokenInformation = new TokenInformation(tokenId, owner,
+            context.principal(), renewers, now, maxTimestamp, expiryTimestamp);
+
+        DelegationTokenData newDelegationTokenData = new DelegationTokenData(newTokenInformation);
+
+        responseData
+                .setErrorCode(NONE.code())
+                .setIssueTimestampMs(now)
+                .setExpiryTimestampMs(expiryTimestamp)
+                .setMaxTimestampMs(maxTimestamp)
+                .setTokenId(tokenId)
+                .setHmac(hmac);
+
+        records.add(new ApiMessageAndVersion(newDelegationTokenData.toRecord(), (short) 0));
+        return ControllerResult.atomicOf(records, responseData);
+    }
+
+    public ControllerResult<RenewDelegationTokenResponseData> renewDelegationToken(
+        ControllerRequestContext context,
+        RenewDelegationTokenRequestData requestData,
+        MetadataVersion metadataVersion
+    ) {
+        long now = time.milliseconds();
+
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+        RenewDelegationTokenResponseData responseData = new RenewDelegationTokenResponseData();
+
+        TokenInformation myTokenInformation = getToken(requestData.hmac());
+
+        if (myTokenInformation == null) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_NOT_FOUND.code()));
+        }
```

Comment:
Is this what happens if delegation tokens are disabled?  Should we check explicitly instead?

Reply:
Yes, I added the check here and in expire